File: sando_py/sando_py/nodes/fake_sim.py
"""Lightweight kinematic simulator — Python port of src/sando/fake_sim.cpp.

Publishes the drone's State (and optionally Odometry) by integrating from the
last Goal received. Uses Hopf-fibration quaternion for attitude (matches C++).
Skips the Gazebo set_entity_state service path — that's only relevant when
running with Gazebo, in which case the C++ node is preferred anyway. The TF
broadcast and visualization-marker pipeline is kept intact.
"""
from __future__ import annotations

import math
from threading import Lock

import numpy as np

import rclpy
from rclpy.executors import MultiThreadedExecutor
from rclpy.node import Node
from rclpy.qos import QoSDurabilityPolicy, QoSHistoryPolicy, QoSProfile, QoSReliabilityPolicy

from geometry_msgs.msg import TransformStamped, Vector3, Quaternion
from nav_msgs.msg import Odometry
from tf2_ros import TransformBroadcaster
from visualization_msgs.msg import Marker


def _state_msg():
    try:
        from dynus_interfaces.msg import State
        return State
    except ImportError:
        return None


def _goal_msg():
    try:
        from dynus_interfaces.msg import Goal
        return Goal
    except ImportError:
        return None


def _qos_reliable() -> QoSProfile:
    return QoSProfile(history=QoSHistoryPolicy.KEEP_LAST, depth=10,
                      reliability=QoSReliabilityPolicy.RELIABLE,
                      durability=QoSDurabilityPolicy.VOLATILE)


def _hopf_quat_xyzw(thrust: np.ndarray, yaw: float) -> np.ndarray:
    """Hopf fibration: build q from (acceleration + gravity) and yaw.
    Returns xyzw quaternion matching the C++ FakeSim.goalCallback.
    """
    g = np.array([thrust[0], thrust[1], thrust[2] + 9.81])
    n = float(np.linalg.norm(g))
    if n < 1e-9:
        return np.array([0.0, 0.0, math.sin(yaw / 2), math.cos(yaw / 2)])
    a, b, c = g / n
    tmp = 1.0 / math.sqrt(2.0 * (1.0 + c))
    qabc = np.array([-b * tmp, a * tmp, 0.0, tmp * (1.0 + c)])
    qpsi = np.array([0.0, 0.0, math.sin(yaw / 2), math.cos(yaw / 2)])
    return _quat_mul(qabc, qpsi)


def _quat_mul(a: np.ndarray, b: np.ndarray) -> np.ndarray:
    ax, ay, az, aw = a
    bx, by, bz, bw = b
    return np.array([
        aw * bx + ax * bw + ay * bz - az * by,
        aw * by - ax * bz + ay * bw + az * bx,
        aw * bz + ax * by - ay * bx + az * bw,
        aw * bw - ax * bx - ay * by - az * bz,
    ])


class FakeSim(Node):
    def __init__(self):
        super().__init__("fake_sim")
        self.get_logger().info("Initializing FakeSim...")

        start_pos = self.declare_parameter("start_pos", [0.0, 0.0, 4.0]).value
        start_yaw = float(self.declare_parameter("start_yaw", -1.57).value)
        self.send_state_to_gazebo = bool(
            self.declare_parameter("send_state_to_gazebo", False).value)
        self.default_goal_z = float(self.declare_parameter("default_goal_z", 0.3).value)
        self.visual_level = int(self.declare_parameter("visual_level", 0).value)
        self.publish_odom = bool(self.declare_parameter("publish_odom", False).value)
        self.odom_topic = str(self.declare_parameter("odom_topic", "odom").value)
        self.odom_frame_id = str(self.declare_parameter("odom_frame_id", "map").value)
        self.base_frame_id = str(self.declare_parameter("base_frame_id", "").value)

        ns = self.get_namespace() or "/"
        self.ns_ = ns.lstrip("/")
        self.target_frame = f"{self.ns_}/base_link"
        if not self.base_frame_id:
            self.base_frame_id = self.target_frame

        StateMsg = _state_msg()
        GoalMsg = _goal_msg()
        if StateMsg is None:
            raise RuntimeError("dynus_interfaces.msg.State not available")

        self.state_pos = np.array([float(start_pos[0]), float(start_pos[1]), float(start_pos[2])])
        self.state_vel = np.zeros(3)
        self.state_quat = _hopf_quat_xyzw(np.zeros(3), start_yaw)

        self.pub_state = self.create_publisher(StateMsg, "state", _qos_reliable())
        self.pub_marker = self.create_publisher(Marker, "drone_marker", _qos_reliable())
        self.pub_odom = self.create_publisher(Odometry, self.odom_topic, _qos_reliable()) \
            if self.publish_odom else None

        if GoalMsg is not None:
            self.sub_goal = self.create_subscription(GoalMsg, "goal", self._goal_cb, 10)
        else:
            self.get_logger().warn("dynus_interfaces.msg.Goal not available — goal subscription disabled")

        self.br = TransformBroadcaster(self, qos=_qos_reliable())
        self.timer = self.create_timer(0.01, self._tick)

        self._lock = Lock()
        self.goal_received = False
        self.goal_stamp = self.get_clock().now()
        self.goal_pos = np.zeros(3)
        self.goal_vel = np.zeros(3)
        self.goal_acc = np.zeros(3)
        self.goal_quat = np.array([0.0, 0.0, 0.0, 1.0])

        self.get_logger().info("FakeSim initialized")

    def _goal_cb(self, msg) -> None:
        thrust = np.array([msg.a.x, msg.a.y, msg.a.z])
        q = _hopf_quat_xyzw(thrust, float(msg.yaw))
        with self._lock:
            self.goal_stamp = self.get_clock().now()
            self.goal_pos = np.array([msg.p.x, msg.p.y, msg.p.z])
            self.goal_vel = np.array([msg.v.x, msg.v.y, msg.v.z])
            self.goal_acc = thrust
            self.goal_quat = q
            self.goal_received = True
            self.state_pos = self.goal_pos.copy()
            self.state_vel = self.goal_vel.copy()
            self.state_quat = q

    def _tick(self) -> None:
        now = self.get_clock().now()
        with self._lock:
            if self.goal_received:
                dt = max(0.0, min(0.1, (now - self.goal_stamp).nanoseconds * 1e-9))
                p_interp = self.goal_pos + self.goal_vel * dt + 0.5 * self.goal_acc * dt * dt
                quat = self.goal_quat
            else:
                p_interp = self.state_pos.copy()
                quat = self.state_quat.copy()

        # TF broadcast
        tfm = TransformStamped()
        tfm.header.stamp = now.to_msg()
        tfm.header.frame_id = "map"
        tfm.child_frame_id = self.target_frame
        tfm.transform.translation.x = float(p_interp[0])
        tfm.transform.translation.y = float(p_interp[1])
        tfm.transform.translation.z = float(p_interp[2])
        tfm.transform.rotation.x = float(quat[0])
        tfm.transform.rotation.y = float(quat[1])
        tfm.transform.rotation.z = float(quat[2])
        tfm.transform.rotation.w = float(quat[3])
        self.br.sendTransform(tfm)

        # State publish
        StateMsg = _state_msg()
        if StateMsg is not None:
            s = StateMsg()
            s.header.stamp = now.to_msg()
            s.header.frame_id = "map"
            s.pos = Vector3(x=float(p_interp[0]), y=float(p_interp[1]), z=float(p_interp[2]))
            s.vel = Vector3(x=float(self.state_vel[0]), y=float(self.state_vel[1]),
                            z=float(self.state_vel[2]))
            s.quat = Quaternion(x=float(quat[0]), y=float(quat[1]),
                                z=float(quat[2]), w=float(quat[3]))
            self.pub_state.publish(s)

        # Odom
        if self.pub_odom is not None:
            odom = Odometry()
            odom.header.stamp = now.to_msg()
            odom.header.frame_id = self.odom_frame_id
            odom.child_frame_id = self.base_frame_id
            odom.pose.pose.position.x = float(p_interp[0])
            odom.pose.pose.position.y = float(p_interp[1])
            odom.pose.pose.position.z = float(p_interp[2])
            odom.pose.pose.orientation.x = float(quat[0])
            odom.pose.pose.orientation.y = float(quat[1])
            odom.pose.pose.orientation.z = float(quat[2])
            odom.pose.pose.orientation.w = float(quat[3])
            odom.twist.twist.linear.x = float(self.state_vel[0])
            odom.twist.twist.linear.y = float(self.state_vel[1])
            odom.twist.twist.linear.z = float(self.state_vel[2])
            self.pub_odom.publish(odom)

        # Drone mesh marker (optional)
        if self.visual_level > 0:
            m = Marker()
            m.id = 1
            m.ns = f"mesh_{self.ns_}"
            m.header.frame_id = "map"
            m.header.stamp = now.to_msg()
            m.type = Marker.MESH_RESOURCE
            m.action = Marker.ADD
            m.pose.position.x = float(p_interp[0])
            m.pose.position.y = float(p_interp[1])
            m.pose.position.z = float(p_interp[2])
            m.pose.orientation.x = float(quat[0])
            m.pose.orientation.y = float(quat[1])
            m.pose.orientation.z = float(quat[2])
            m.pose.orientation.w = float(quat[3])
            m.mesh_use_embedded_materials = True
            m.mesh_resource = "package://sando/meshes/quadrotor/quadrotor.dae"
            m.scale.x = m.scale.y = m.scale.z = 0.75
            self.pub_marker.publish(m)


def main(args=None):
    rclpy.init(args=args)
    node = FakeSim()
    executor = MultiThreadedExecutor()
    executor.add_node(node)
    try:
        executor.spin()
    finally:
        executor.shutdown()
        node.destroy_node()
        rclpy.shutdown()


if __name__ == "__main__":
    main()

File: cpp/ros2/src/sando_cpp/src/fake_sim.cpp
// fake_sim.cpp — C++ ROS2 port of sando_py/nodes/fake_sim.py (kinematic loop-closer).
// Subscribes Goal, ideal-follows it (state := goal), publishes State (+ TF + optional Odom)
// at 100Hz with a constant-accel extrapolation. Hopf-fibration attitude (matches the C++ baseline).
#include <rclcpp/rclcpp.hpp>
#include <geometry_msgs/msg/transform_stamped.hpp>
#include <nav_msgs/msg/odometry.hpp>
#include <tf2_ros/transform_broadcaster.h>
#include <dynus_interfaces/msg/state.hpp>
#include <dynus_interfaces/msg/goal.hpp>

#include <array>
#include <cmath>
#include <mutex>
#include <chrono>
using namespace std::chrono_literals;

static std::array<double, 4> quat_mul(const std::array<double, 4>& a, const std::array<double, 4>& b) {
  double ax = a[0], ay = a[1], az = a[2], aw = a[3];
  double bx = b[0], by = b[1], bz = b[2], bw = b[3];
  return {aw * bx + ax * bw + ay * bz - az * by,
          aw * by - ax * bz + ay * bw + az * bx,
          aw * bz + ax * by - ay * bx + az * bw,
          aw * bw - ax * bx - ay * by - az * bz};
}
static std::array<double, 4> hopf_quat(double tx, double ty, double tz, double yaw) {
  double gx = tx, gy = ty, gz = tz + 9.81;
  double n = std::sqrt(gx * gx + gy * gy + gz * gz);
  if (n < 1e-9) return {0.0, 0.0, std::sin(yaw / 2), std::cos(yaw / 2)};
  double a = gx / n, b = gy / n, c = gz / n;
  double tmp = 1.0 / std::sqrt(2.0 * (1.0 + c));
  std::array<double, 4> qabc = {-b * tmp, a * tmp, 0.0, tmp * (1.0 + c)};
  std::array<double, 4> qpsi = {0.0, 0.0, std::sin(yaw / 2), std::cos(yaw / 2)};
  return quat_mul(qabc, qpsi);
}

class FakeSim : public rclcpp::Node {
 public:
  FakeSim() : rclcpp::Node("fake_sim") {
    auto sp = declare_parameter<std::vector<double>>("start_pos", {0.0, 0.0, 4.0});
    double start_yaw = declare_parameter<double>("start_yaw", -1.57);
    publish_odom_ = declare_parameter<bool>("publish_odom", false);
    odom_topic_ = declare_parameter<std::string>("odom_topic", "odom");
    std::string ns = get_namespace(); if (!ns.empty() && ns[0] == '/') ns = ns.substr(1);
    target_frame_ = ns + "/base_link";

    state_pos_ = {sp[0], sp[1], sp[2]}; state_vel_ = {0, 0, 0};
    state_quat_ = hopf_quat(0, 0, 0, start_yaw);
    goal_pos_ = {0, 0, 0}; goal_vel_ = {0, 0, 0}; goal_acc_ = {0, 0, 0};
    goal_quat_ = {0, 0, 0, 1};

    pub_state_ = create_publisher<dynus_interfaces::msg::State>("state", 10);
    if (publish_odom_) pub_odom_ = create_publisher<nav_msgs::msg::Odometry>(odom_topic_, 10);
    sub_goal_ = create_subscription<dynus_interfaces::msg::Goal>(
        "goal", 10, std::bind(&FakeSim::goal_cb, this, std::placeholders::_1));
    br_ = std::make_unique<tf2_ros::TransformBroadcaster>(*this);
    timer_ = create_wall_timer(10ms, std::bind(&FakeSim::tick, this));
    RCLCPP_INFO(get_logger(), "FakeSim (C++ port) initialized");
  }

 private:
  void goal_cb(const dynus_interfaces::msg::Goal::SharedPtr msg) {
    auto q = hopf_quat(msg->a.x, msg->a.y, msg->a.z, msg->yaw);
    std::lock_guard<std::mutex> lk(mtx_);
    goal_stamp_ = now();
    goal_pos_ = {msg->p.x, msg->p.y, msg->p.z};
    goal_vel_ = {msg->v.x, msg->v.y, msg->v.z};
    goal_acc_ = {msg->a.x, msg->a.y, msg->a.z};
    goal_quat_ = q;
    goal_received_ = true;
    state_pos_ = goal_pos_; state_vel_ = goal_vel_; state_quat_ = q;
  }

  void tick() {
    rclcpp::Time t = now();
    std::array<double, 3> p; std::array<double, 4> quat;
    {
      std::lock_guard<std::mutex> lk(mtx_);
      if (goal_received_) {
        double dt = std::max(0.0, std::min(0.1, (t - goal_stamp_).seconds()));
        for (int i = 0; i < 3; ++i) p[i] = goal_pos_[i] + goal_vel_[i] * dt + 0.5 * goal_acc_[i] * dt * dt;
        quat = goal_quat_;
      } else { p = state_pos_; quat = state_quat_; }
    }
    geometry_msgs::msg::TransformStamped tf;
    tf.header.stamp = t; tf.header.frame_id = "map"; tf.child_frame_id = target_frame_;
    tf.transform.translation.x = p[0]; tf.transform.translation.y = p[1]; tf.transform.translation.z = p[2];
    tf.transform.rotation.x = quat[0]; tf.transform.rotation.y = quat[1];
    tf.transform.rotation.z = quat[2]; tf.transform.rotation.w = quat[3];
    br_->sendTransform(tf);

    dynus_interfaces::msg::State s;
    s.header.stamp = t; s.header.frame_id = "map";
    s.pos.x = p[0]; s.pos.y = p[1]; s.pos.z = p[2];
    s.vel.x = state_vel_[0]; s.vel.y = state_vel_[1]; s.vel.z = state_vel_[2];
    s.quat.x = quat[0]; s.quat.y = quat[1]; s.quat.z = quat[2]; s.quat.w = quat[3];
    pub_state_->publish(s);

    if (pub_odom_) {
      nav_msgs::msg::Odometry odom;
      odom.header.stamp = t; odom.header.frame_id = "map"; odom.child_frame_id = target_frame_;
      odom.pose.pose.position.x = p[0]; odom.pose.pose.position.y = p[1]; odom.pose.pose.position.z = p[2];
      odom.pose.pose.orientation.x = quat[0]; odom.pose.pose.orientation.y = quat[1];
      odom.pose.pose.orientation.z = quat[2]; odom.pose.pose.orientation.w = quat[3];
      odom.twist.twist.linear.x = state_vel_[0]; odom.twist.twist.linear.y = state_vel_[1];
      odom.twist.twist.linear.z = state_vel_[2];
      pub_odom_->publish(odom);
    }
  }

  std::mutex mtx_;
  bool goal_received_ = false, publish_odom_ = false;
  std::string odom_topic_, target_frame_;
  std::array<double, 3> state_pos_, state_vel_, goal_pos_, goal_vel_, goal_acc_;
  std::array<double, 4> state_quat_, goal_quat_;
  rclcpp::Time goal_stamp_;
  rclcpp::Publisher<dynus_interfaces::msg::State>::SharedPtr pub_state_;
  rclcpp::Publisher<nav_msgs::msg::Odometry>::SharedPtr pub_odom_;
  rclcpp::Subscription<dynus_interfaces::msg::Goal>::SharedPtr sub_goal_;
  std::unique_ptr<tf2_ros::TransformBroadcaster> br_;
  rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char** argv) {
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<FakeSim>());
  rclcpp::shutdown();
  return 0;
}
